streams/src/main/java/org/apache/kafka/streams/processor/internals/SinkNode.java

```diff
@@ -57,7 +57,7 @@ public void init(ProcessorContext context) {
     public void process(K key, V value) {
         // send to all the registered topics
         RecordCollector collector = ((RecordCollector.Supplier) context).recordCollector();
-        collector.send(new ProducerRecord<>(topic, key, value), keySerializer, valSerializer, partitioner);
+        collector.send(new ProducerRecord<>(topic, null, context.timestamp(), key, value), keySerializer, valSerializer, partitioner);
```

Comment:
I agree that we cannot / do not need to define constructors with all possible combinations, while for this case I think it is more common that partition-ids and timestamps are omitted than key and value. I think it will be common for KafkaProducer in general beyond its usage for Kafka Streams, and hence it is better to add it now than wait for it to be added and possibly come back (if we still remember) and change the APIs.

More generally, I feel it is better to not specifying nulls in the APIs but let the class to handle sentinels (because sometimes nulls may not be the appropriate sentinel as well).